### PR TITLE
add a helper class for runtime-derived uint32

### PR DESCRIPTION
**Pull request by WeavingGao**
Signed-off-by: gaoweiwen <whereisgww@outlook.com>

Envoy already has several runtime protobuf messages, such as `RuntimeUInt32`, `RuntimePercent` and `RuntimeDouble`. Each protobuf message has a corresponding C++ helper class to easily and safely obtain the runtime value, except for `RuntimeUInt32`. So this PR tries to add the missing helper class.

TODO(WeavingGao): use for #16392

Commit Message: helper class for `RuntimeUInt32`
Additional Description:
Risk Level: Low
Testing: Unit Test
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by WeavingGao** [comment]
ready to be reviewed.

File: test/common/runtime/runtime_protos_test.cc
Comment (antoniovicente) [inline test/common/runtime/runtime_protos_test.cc:43]: In cases where getInteger returns a value larger than uint32 max, should this mask the returned value down to 32 bits or return the default value?  Consistency between the behavior of getDouble and getInteger for values larger than uint64 max suggest that we should return the default in this case.

https://github.com/envoyproxy/envoy/blob/main/source/common/runtime/runtime_impl.h#L104

**Review by antoniovicente — COMMENTED** [review]
cc @tonya11en 

Hey Tony, could you do a review pass on this code since you were involved in the implementation of the wrapper for doubles?

File: test/common/runtime/runtime_protos_test.cc
Comment (WeavingGao) [reply test/common/runtime/runtime_protos_test.cc:43]: The default value is indeed more reasonable, I'll submit a commit to fix it.

Comment (WeavingGao) [reply test/common/runtime/runtime_protos_test.cc]: Fixed in the latest commit. @antoniovicente

**Review by tonya11en — CHANGES_REQUESTED** [review]
Just a minor comment, LG otherwise. Thanks!

File: source/common/runtime/runtime_protos.h
Comment (WeavingGao) [reply source/common/runtime/runtime_protos.h:25]: This may introduce additional complexity. If we introduce additional variables to record the number of invalid parsing, this class will become thread-unsafe, making it either complicated or poor in performance. 
I'm not sure if there are other solutions, I look forward to your suggestions.

Comment (tonya11en) [reply source/common/runtime/runtime_protos.h:25]: I mean just put ENVOY_LOG_EVERY_POW_2()  or ENVOY_LOG_EVERY_N() in a conditional instead of using that ternary operator. You don't need to get too clever here. 

I would be genuinely surprised if that caused an appreciable decline in performance.

File: source/common/runtime/runtime_protos.h
Comment (tonya11en) [inline source/common/runtime/runtime_protos.h:27]: Sorry one more-- can you mention in the message that it's a parsed runtime value and swap "return" for "returning". 

It'll be good to go after that.

**Review by tonya11en — CHANGES_REQUESTED** [review]

File: source/common/runtime/runtime_protos.h
Comment (WeavingGao) [reply source/common/runtime/runtime_protos.h]: I misunderstood what you meant before (embarrassing >_<). Now I understand your suggestion and submit a new commit.

File: source/common/runtime/runtime_protos.h
Comment (WeavingGao) [reply source/common/runtime/runtime_protos.h:27]: Sure.

**Review by tonya11en — DISMISSED** [review]
Thanks!

**Comment by WeavingGao** [comment]
Could you please take a look since you had implemented a similar wrapper class and also have write access to merge. @alyssawilk

**Review by alyssawilk — COMMENTED** [review]
I'm happy to take a look, but I'm unclear on why we're landing this upstream if it's not being used.  Maybe if you filled out the Pr description fields it would explain?  I think largely if its not used in Envoy it's going to end up being cleaned up at some point.

**Comment by WeavingGao** [comment]
> I'm happy to take a look, but I'm unclear on why we're landing this upstream if it's not being used. Maybe if you filled out the Pr description fields it would explain? I think largely if its not used in Envoy it's going to end up being cleaned up at some point.

When I tried to implement another feature (see #16392 ), I found that I need a `RuntimeUInt32` helper class just like `RuntimeDouble`. I was going to implement the helper class in that PR, but I thought it was very general, so I submitted it in a separate PR. The helper class is not used currently, but the `RuntimeUInt32` protobuf message does exists and any other place that uses this protobuf message can benefit from the helper class.

**Comment by tonya11en** [comment]
@WeavingGao can you please update the description to include those details. This way other's don't have to fish through the comments to gather context.

**Comment by WeavingGao** [comment]
> @WeavingGao can you please update the description to include those details. This way other's don't have to fish through the comments to gather context.

Give some details in the PR description. @alyssawilk @tonya11en

File: source/common/runtime/runtime_protos.h
Comment (alyssawilk) [inline source/common/runtime/runtime_protos.h:21]: Mind adding a unit test for this?

**Review by alyssawilk — COMMENTED** [review]
Gotcha, thanks for the context.
I'd mildly prefer this land with the actual use in #16392, but if you'd prefer to land it separately how about a TODO(WeavingGao): use for #16392 just to make it clear it's not used yet, but what the plan is?

File: source/common/runtime/runtime_protos.h
Comment (WeavingGao) [reply source/common/runtime/runtime_protos.h:21]: done.

**Review by alyssawilk — COMMENTED** [review]
lgtm modulo one nit - could you put the TODO in the code, rather than the PR description?  Sorry that wasn't clear!

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**
